src/Functions/initializeAggregation.cpp

```diff
@@ -0,0 +1,161 @@
+
+        AggregateFunctionProperties properties;
+        aggregate_function = AggregateFunctionFactory::instance().get(aggregate_function_name, argument_types, params_row, properties);
+    }
+
+    return aggregate_function->getReturnType();
+}
+
+
+void FunctionInitializeAggregation::executeImpl(Block & block, const ColumnNumbers & arguments, size_t result, size_t input_rows_count)
+{
+    IAggregateFunction & agg_func = *aggregate_function;
+    std::unique_ptr<Arena> arena = std::make_unique<Arena>();
+
+    const size_t num_arguments_columns = arguments.size() - 1;
+
+    std::vector<ColumnPtr> materialized_columns(num_arguments_columns);
+    std::vector<const IColumn *> aggregate_arguments_vec(num_arguments_columns);
+
+    for (size_t i = 0; i < num_arguments_columns; ++i)
+    {
+        const IColumn * col = block.getByPosition(arguments[i + 1]).column.get();
+        materialized_columns.emplace_back(col->convertToFullColumnIfConst());
+        aggregate_arguments_vec[i] = &(*materialized_columns.back());
+    }
+
+    const IColumn ** aggregate_arguments = aggregate_arguments_vec.data();
+
+    MutableColumnPtr result_holder = block.getByPosition(result).type->createColumn();
+    IColumn & res_col = *result_holder;
+
+    /// AggregateFunction's states should be inserted into column using specific way
+    auto * res_col_aggregate_function = typeid_cast<ColumnAggregateFunction *>(&res_col);
+
+    if (!res_col_aggregate_function && agg_func.isState())
+        throw Exception("State function " + agg_func.getName() + " inserts results into non-state column "
+                        + block.getByPosition(result).type->getName(), ErrorCodes::ILLEGAL_COLUMN);
+
+    PODArray<AggregateDataPtr> places(input_rows_count);
+    for (size_t i = 0; i < input_rows_count; ++i)
+    {
+        places[i] = arena->alignedAlloc(agg_func.sizeOfData(), agg_func.alignOfData());
+        try
+        {
+            agg_func.create(places[i]);
+        }
+        catch (...)
+        {
+            for (size_t j = 0; j < i; ++j)
+                agg_func.destroy(places[j]);
+            throw;
+        }
+    }
+
+    SCOPE_EXIT({
+        for (size_t i = 0; i < input_rows_count; ++i)
+            agg_func.destroy(places[i]);
+    });
+
+    {
+        auto * that = &agg_func;
+        /// Unnest consecutive trailing -State combinators
+        while (auto * func = typeid_cast<AggregateFunctionState *>(that))
+            that = func->getNestedFunction().get();
+        that->addBatch(input_rows_count, places.data(), 0, aggregate_arguments, arena.get());
+    }
+
+    for (size_t i = 0; i < input_rows_count; ++i)
+        if (!res_col_aggregate_function)
+            agg_func.insertResultInto(places[i], res_col, arena.get());
+        else
+            res_col_aggregate_function->insertFrom(places[i]);
+    block.getByPosition(result).column = std::move(result_holder);
```

Comment:
If it's the exact code from `arrayReduce` function, maybe move it to seperate funtion with name like `applyAggregation`.

Reply:
not the exact same code, we accept different kind of arguments and we have different function call to execute the aggregation, does it's worth it ?

Reply:
I see, `addBatch` vs `addBatchArray`. Ok, let leave it as is.